trlx/trainer/accelerate_base_trainer.py

```diff
@@ -503,17 +503,31 @@ def learn(self):  # noqa: C901
                         results = self.evaluate()
                         stats.update(results)
 
-                        # FIXME: seems to not work with zero and barriers don't seem to help
-                        if (
-                            self.config.train.save_best
-                            and int(os.environ.get("DEEPSPEED_ZERO_STAGE", -1)) == -1
-                        ):
-                            if (
-                                "reward/mean" in stats
-                                and stats["reward/mean"] > best_reward
+                        # always save checkpoint with the greatest mean reward
+                        if self.config.train.save_best:
+                            if stats.get("reward/mean", -float("inf")) > best_reward:
+                                best_reward = stats.get("reward/mean")
+                                do_save = True
+                            # in case ILQL reports reward estimate as one of its metrics
+                            elif (
+                                stats.get("metrics/reward", -float("inf")) > best_reward
                             ):
-                                best_reward = stats["reward/mean"]
-                                self.save("best_checkpoint")
+                                best_reward = stats.get("metrics/reward")
+                                do_save = True
+                            else:
+                                do_save = False
+                            do_save = torch.tensor(
+                                do_save, device=self.accelerator.device
+                            )
+                            torch.distributed.all_reduce(
+                                do_save, torch.distributed.ReduceOp.MAX
+                            )
+                            if do_save:
+                                best_path = f"{self.config.train.checkpoint_dir}/best_checkpoint"
```

Comment:
Nice - forgot to change this when updating the behavior to checkpoint into the specified dir.